### ISSUE TITLE
Support foreground with fail2ban-server

### ISSUE THREAD

**Issue by Glandos** [fail2ban/fail2ban#1139]
With the era of new-style init (such as systemd) or daemon managers (such as supervisord), it seems safer to write program that doesn't try to be a daemon itself, since there are a lot of corner cases and pitfalls. Fail2ban seems to do it nicely, however, I see at least two reasons to support a foreground mode:
  * If some platform has some specific initialization pattern, it could be handled by the init/daemon manager.
  * Log management is far easier when the init/daemon manager handles it, usually by fetching the stdout of the launched program. I'm not sure that setting the logtarget to STDOUT with the current daemonizing behavior will be successful.

Is there any good reason not to support a foreground behavior? Or is it a simple historical thing?

**Comment by yarikoptic** [comment]
well -- the fail2ban-server itself already can be launched in foreground: -f option.  Actually even client has -f but not sure what it is doing really because I thought it would do exactly that -- start server, feeds it with  configs and then wait for that server process to die off...  So someone might just need to look into it.

**Comment by Glandos** [comment]
According to https://github.com/fail2ban/fail2ban/blob/master/bin/fail2ban-client#L291 the client fork itself, before starting the server, whatever the mode is.
I don't know if the server can be launched with the right command-line (see the execv right after in the same method). Because right now, if I understand it correctly (see https://github.com/fail2ban/fail2ban/blob/master/bin/fail2ban-client#L276) the configuration is parsed by fail2ban-client and streamed to the server.

**Comment by sebres** [comment]
It's a historical thing... To start resp. run it completely as server, we should move the part of the client functionality, that read config files, to the server (or better to new thread in server, something like ConfigReaderThread, that can read/reload/stop/etc and ends hereafter).

In any case, the good point would be to make it together with #1121

**Comment by yarikoptic** [comment]
On Wed, 29 Jul 2015, Serg G. Brester wrote:

>    It's a historical thing... To start resp. run it completely as server, we
>    should move the part of the client functionality, that read config files,
>    to the server (or better to new thread in server, something like
>    ConfigReaderThread, that can read/reload/stop/etc and ends hereafter).

>    In any case, the good point would be to make it together with #1121

better not. #1121 is just a mere refactoring moving code around and
providing entry points, which shouldn't affect any functionality
(will do right after 0.9.3 release).  Fix for this one would require
functionality changes, so better be done separately after that.

and I thought that it could be done with minimal blood without
moving/sharing functionality between client/server, just letting client
process start server separately, do its "initiation" dance and
then joining the server process (thus remaining in foreground). Didn't
have a chance to try.  As such this change would even remain orthogonal
(independent) of #1121.

**Comment by sebres** [comment]
Agree, or in fact, as you said, after #1121 will be ready. My point was, it should not be done before (just makes few sense).

**Comment by jsumners** [comment]
This is still an issue. I'm trying to provide a solution for it in https://github.com/voidlinux/void-packages/issues/3545

First,  `fail2ban-server` doesn't parse the config file and uses hard coded paths for `socket` and `pidfile`, which it won't create if `/var/run/fail2ban/` doesn't already exist. See:

* https://github.com/fail2ban/fail2ban/blob/master/bin/fail2ban-server#L52
* https://github.com/fail2ban/fail2ban/blob/master/fail2ban/server/server.py#L106

Second, `fail2ban-client -f`, as has been said in this thread, doesn't really launch the server in the foreground. And simply launching the server via the client and then doing `exec pause` is insufficient. Under runit, this will cause the fail2ban server process to start outside of the runsvdir process. Thus, when stopping the process, `sv stop fail2ban` (or whatever), you don't really stop the server.

Personally, I think the solution is to make the `fail2ban-server` binary script read the configuration. It's foreground launch really does work.

**Comment by sebres** [comment]
I've a major private branch, where the server reads a config directly (reading of the config functionality was moved from the client), but it is based on my another branch, that is incompatible with upstream master.

I doubt, that I could do a rebase on current master here. But I will try it tomorrow...

**Comment by sebres** [comment]
So, I've rebased the mentioned CS-changes to my major public branch as new sub-branch `f2b-perfom-prepare-716-cs`:
https://github.com/sebres/fail2ban/tree/f2b-perfom-prepare-716-cs
At the moment It's too hard to merge/rebase it relative current master, because too much different, but I'll see, what can I do.
I would ask all involved (and @fail2ban/contributors) to test this, as it is (in this branch).

To see all the changes, follow this link:
https://github.com/sebres/fail2ban/compare/f2b-perfom-prepare-716...sebres:f2b-perfom-prepare-716-cs

**New are:**
- `fail2ban-server` can be used to start server directly (internally uses client-thread for configure self) in both modes also - background and foreground.
- `fail2ban-client` can work also in both modes, where in background mode started a new fork of server (not from itself);
 - client/server functionality moved away from `bin` and using now the common interface (introduced in fail2bancmdline);
- start in foreground fixed;
- server can act as client corresponding command line;
- command "restart" added: in opposite to "reload" in reality restarts the server (new process);
- several client/server bugs during starting process fixed.

**Known (TODO):**
- several test cases are missing (folder bin resp. client/server was previously at all not covered)

@fail2ban/developers Any suggestions are welcome (but I will had not many time to completely "rewrite" it again (greetings to @yarikoptic :=).

**Comment by sebres** [comment]
Oh something forgotten:
The new solution is much more faster as current master, but:
If someone want to test this new solution by means of test-cases, you can do that with new argument `-f` or `--fast` also (for all of new arguments please call `fail2ban-testcases -h`)...

It's really very-very fast:
```
$ bin/fail2ban-testcases -fn
...
Ran 326 tests in 2.400s
```
Compared to the testing of the current master, I need on same computer about 120 seconds.

**Comment by yarikoptic** [comment]
@sebres could you just  send a PR with whatever you think should be merged?

**Comment by sebres** [comment]
@yarikoptic well, as I already wrote 2 things:
- "ATM It's too hard to merge/rebase it relative current master", I made it relative my branch (see comparison above).
- The PR will decrease coverage to 86% (see TODO above)...

**Comment by yarikoptic** [comment]
And it will keep getting only harder, not easier unless piece by piece we merge it all in.  Is there a chance to get at least one feature as a PR to be reviewed independently from the rest of everything?

**Comment by szepeviktor** [comment]
@sebres How can `--fast` really be that much faster?

**Comment by sebres** [comment]
@szepeviktor Many things cause that, for some details see #1116 (previous branch).
But this solution is not only in the test cases faster, I saw in production at least 2-10 times faster (and also less CPU and memory hungry).

**Comment by jsumners** [comment]
So is there a consensus on this?

**Comment by sebres** [comment]
I work still on the test cases for the new PR. So comming soon.
But that will be first a killer PR, and I've no time to rebase it (or somehow partially merge it) to the current master...

**Comment by sebres** [comment]
@yarikoptic see #1321 (as already said - killer PR, but if someone can it better - welcome;o)

**Comment by sebres** [comment]
done in 0.10-branch per #1413
